### ISSUE TITLE
PlayCookieSessionStore: can't save custom class due to "untrusted package error"

### ISSUE THREAD

**Issue by unkarjedy** [pac4j/play-pac4j#252]
Yesterday I tried to move from `PlayCacheSessionStore` to `PlayCookieSessionStore` in order to make our server session-stateless.  Our session contais quite a lightweight user info as a instance of case class. It is saved via  `PlayCookieSessionStore.set` with key `pac4j_pac4jUserProfiles`.
But unfortunately `null` is returned when `PlayCookieSessionStore.get` is called for the same key.

Minimum Complete and Verifiable Example:
```scala
package _scala_interestings

import org.pac4j.play.PlayWebContext
import org.pac4j.play.store.PlayCookieSessionStore
import play.core.j.JavaHelpers
import play.mvc.Http
import play.mvc.Http.RequestBuilder

object PlayCookieSessionStoreTest {

  case class User(firstName: String, lastName: String)

  def main(args: Array[String]): Unit = {
    val sessionStore = new PlayCookieSessionStore()
    val httpContext = new Http.Context(new RequestBuilder(), JavaHelpers.createContextComponents())
    val context = new PlayWebContext(httpContext, sessionStore)

    println("Test 1: set/get string. Everything is OK")
    val key1 = "key1"
    val string = "just simple string"
    sessionStore.set(context, key1, string)
    println(sessionStore.get(context, key1))

    println("Test 2: set/get string. Retrieved value is NULL!")
    val key2 = "key2"
    val profile = User("firstName", "lastName")
    sessionStore.set(context, key2, profile)
    println(sessionStore.get(context, key2))
  }

}
```

I found the root of the issue. `PlayCookiesessionStore` has following field:
```java
 public static final JavaSerializationHelper JAVA_SER_HELPER = new JavaSerializationHelper();
```
Method `JavaSerializationHelper.unserializeFromBytes` is used inside `PlayCookiesessionStore.get`. Inside that method a special list is used:
```java
private List<String> trustedPackages;
```
that is initialized in the following way:
```java
    public JavaSerializationHelper() {
        trustedPackages = new ArrayList<>();
        trustedPackages.addAll(Arrays.asList("java.", "javax.", "org.pac4j.", "com.github.scribejava.", "org.opensaml.", "com.nimbusds.",
            "org.joda.", "net.minidev.json.", "org.bson.types."));
    }
```
The list of trusted packages does not contain our app package, thus it does not work as we expect.

**Comment by unkarjedy** [comment]
@leleuj 

To bypass the issue we have to write something like this:
```scala
PlayCookieSessionStore.JAVA_SER_HELPER.getTrustedPackages.addAll(Seq("org.myapp").asJava)
or
PlayCookieSessionStore.JAVA_SER_HELPER.setTrustedPackages(PlayCookieSessionStore.JAVA_SER_HELPER.getTrustedPackages)
```
which IMO is quite a clumsy solution due to:
1) we deal with a global mutable state
2) it is really non-intuitive what is root of the issue, at least we should add some notes to documentation

**Comment by unkarjedy** [comment]
My first/quick proposal is:
1) At least add some `addTrustedPackages` method to `JavaSerializationHelper`
2) Make `trustedPackages` immutable (do not leak from `getTrustedPackages` method, return the copy of the list)
3) Make `trustedPackages` a `Set` not a `List`
4) Make `JAVA_SER_HELPER` private non-static instance
5) Add `addTrustedPackages` to `PlayCookieSessionStore` that will delegate it to `JAVA_SER_HELPER`. `JAVA_SER_HELPER` is a helper and an implementation detail
6) In addition to `trustedPackages` add `trustedClasses` simmilary. Such way we will be able to add trusted classes exmplicitly, not knowing their name, like `addTrustedClass(classOf[User])`
7) Add appropriate notes to documentation

@leleuj When we agree with some solution I will be able to prepare PR.

**Comment by unkarjedy** [comment]
Also, there are some issues with classloaders. If we add customer class ` RestrictedObjectInputStream` will not be able to resolve it due to it uses `play.runsupport.NamedURLClassLoader` with name `DependencyClassLoader` that can't resolve custom classes. (pack4j is added as a jar via sbt libraryDependencies).

**Comment by unkarjedy** [comment]
We probably should not use parent `super.resolveClass` inside `RestrictedObjectInputStream`. Under the hood it uses `latestUserDefinedLoader()` that returns `DependencyClassLoader`.
Maybe we should be able to pass `classLoader` parameter to `RestrictedObjectInputStream` and use it during resolve?
- [Apache Commons IO ClassLoaderObjectInputStream.java](https://github.com/apache/commons-io/blob/master/src/main/java/org/apache/commons/io/input/ClassLoaderObjectInputStream.java)
- [StackOverflow deserialize dynamically loaded class java
](https://stackoverflow.com/questions/35013780/deserialize-dynamically-loaded-class-java)

**Comment by leleuj** [comment]
You're right, there is room for improvement here. Let's take things in order...

**Comment by leleuj** [comment]
About the trusted packages, I'm globally in line with the proposed changes, except that we can't break backward compatibility. So we must keep the list instead of a set in the 3.5.x version and return a mutable list, but we should mark the getter and setter methods with `@Deprecated` and add a new `addTrustedPackages` method.

In the version 4.0.x, we'll be able to change that and use a set and return an immutable set in a getter.

Let's prepare a first PR for the current master (3.5.x) in that way.

**Comment by unkarjedy** [comment]
@leleuj Ok, I will prepare a PR this weak

**Comment by unkarjedy** [comment]
@leleuj 
Her is the first PR: https://github.com/pac4j/pac4j/pull/1225
(Did I said this weak? seems like I meant this year =D)

**Comment by leleuj** [comment]
@unkarjedy Thanks. Just reviewed it. LGTM. As pac4j v4 has started, we now need two PRs: one (backward compatible) for the 3.5.x branch and one for the master...

**Comment by leleuj** [comment]
Fixed in 3.5.0 and 4.0.0...
